### PR TITLE
feat(devtools-example): Visualize Shared-Tree in Devtools

### PR DESCRIPTION
**Pull request by jikim-msft**
#### Description

https://dev.azure.com/fluidframework/internal/_workitems/edit/4495/

This PR visualizes `@fluid-experimental/tree2` package's `SharedTree`.

#### TODO
~~Find API which retrieves the schema in JSON-format and use it for `visualizeSharedTree`.~~

#### After

Screenshot 2023-06-09 at 8 46 56 AM

### CODE REVIEW

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts [hunk 1]
```diff
@@ -25,6 +25,16 @@ import {
 	VisualTreeNode,
 } from "./VisualTree";
 
+/**
+ * @public
+ */
+// eslint-disable-next-line @typescript-eslint/no-extraneous-class
+export class SharedTreeObject {
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:32]: I don't think we want to export this here. If this factory object isn't available in SharedTree's API, we should probably add it. This is a core pattern available in other DDS packages. @noencke is there any reason we can't expose the factory statics in SharedTree?

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts [hunk 2]
```diff
@@ -190,6 +200,34 @@ export const visualizeSharedString: VisualizeSharedObject = async (
...
+	visualizeChildData: VisualizeChildData,
+): Promise<FluidObjectTreeNode> => {
+	const sharedTree = sharedObject as ISharedTree;
+
+	console.log("Default Visualizer :", JSON.stringify(sharedTree.root));
+
+	// TODO: Find Shared Tree API for JSON data format.
+	const children: Record<string, VisualChildNode> = {};
+	if (sharedTree.root !== undefined) {
+		for (const [key, value] of Object.entries(sharedTree.root)) {
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:217]: Is there a more type-safe API we can use to walk the tree's children? Or is this the canonical way to walk a SharedTree's contents (when you don't have a schema)?

File: packages/tools/devtools/devtools-example/src/App.tsx [hunk 3]
```diff
@@ -104,6 +161,12 @@ async function populateRootMap(container: IFluidContainer): Promise<void> {
 			b: "b",
 		},
 	});
+
+	// Add some unrecognizable data to test devtools handling
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/App.tsx:165]: Looks like this got re-added. I removed these as a fix for the infinite spinner issue we were seeing in the data view.

File: packages/tools/devtools/devtools-example/package.json [hunk 4]
```diff
@@ -40,6 +40,7 @@
 		"webpack:dev": "webpack --env development"
 	},
 	"dependencies": {
+		"@fluidframework/aqueduct": "workspace",
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/package.json:43]: Doesn't look like this one is being used.

File: packages/tools/devtools/devtools-example/src/App.tsx [hunk 5]
```diff
@@ -93,6 +123,33 @@ async function populateRootMap(container: IFluidContainer): Promise<void> {
...
+	);
+
+	const view = sharedTree.schematize({
+		schema,
+		allowedSchemaModifications: AllowedUpdateType.None,
+		initialTree: { children: ["hello", "world"] },
+	});
+
+	// TODO: Console to avoid linter problem. Make sure to remove
+	console.log("Live Forever:", view.root);
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/App.tsx:149]: TODO :)

Comment (jikim-msft) [reply packages/tools/devtools/devtools-example/src/App.tsx:149]: That's embarrassing! Just in case you were curious, it was the song I was listening to when I was working on it 😅.

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts (hunk 1 above)
Comment (noencke) [reply packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:32]: Agreed it probably shouldn't be exported. 

The `container.create([DDS_Class])` function is oddly specific in that it requires not just a static factory method on the given class object but also the class object must have a constructor which returns a type with a `handle` field. As far as we have seen, that constructor is never actually invoked. The API typing seems to be attempting to force the user to provide the DDS class object rather than, say, just asking for a function which produces instances of that class. I'm not sure why it is like this; perhaps it is trying to keep people from making mistakes using the API? The actual class `SharedTree` is intentionally not exposed; we provide an interface `ISharedTree` that reduces the API surface to what we want (otherwise, all kinds of types get exposed which should stay internal). 

We anticipate that going forward, the API for consuming these factories will need to change eventually for an additional reason: they don't allow for passing arguments to the DDS creation. SharedTree is going to need that, and really any DDS could/should (e.g. passing your DDS a "summary version" after you've done a data format migration). Without having that, even if we exposed the SharedTree class, apps would have to make subclasses of SharedTree anyway which overwrite the kind of _factory_ that they expose, where that factory is hardcoded to provide the config that is passed into the trees at construction time. This is what some customers of the legacy SharedTree had to do.

I think this is a conversation that needs to continue so as to inform the evolution of the initialization APIs. In the meantime, since both "make this little wrapper class and do a cast" and "expose the SharedTree class" are temporary workarounds, we opted for the former since it's much, much easier.

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:215]: Assuming you want a simple readonly view there are a couple of approaches.

1. If you want the verbose view where value fields (always exactly one child) arn't inlined and such, tree.context.root to get the root field. Use .length and getNode(index) to read the nodes. Iterate the node to get the fields from it, and use the [valueSymbol]  and [typeNameSymbol] to get the rest of the data from the node.
2. If you want instead the more concise version with primitive nodes replaces with just primitive types and value fields inclined and such, tree.root to get the root. Use isEditiableField to tell if you got a field or node. If field, iterate it to get its children. If node get the field gets with object.onwKeys I think it is, then read those fields off the object with getField symbol.

Comment (CraigMacomber) [reply packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:215]: Also note that technically you always have a schema (one is stored in the document, you can read it via the storedSchema field. What you don't have is a statically known schema.

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts (hunk 1 above)
Comment (alexvy86) [reply packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:32]: cc @Abe27342 since I know you've mentioned being able to pass arguments to DDSes.

File: packages/tools/devtools/devtools-example/src/App.tsx [hunk 6]
```diff
@@ -93,6 +111,30 @@ async function populateRootMap(container: IFluidContainer): Promise<void> {
 	const sharedCounter = await container.create(SharedCounter);
 	rootMap.set(sharedCounterKey, sharedCounter.handle);
 
+	// Set up SharedTree for visualization
+	const sharedTreeFactoryObject = SharedTreeObject as unknown as SharedObjectClass<ISharedTree>;
+	const sharedTree = await container.create(sharedTreeFactoryObject);
+
+	// Populate SharedTree using Schema
+	const builder = new SchemaBuilder("SchemaAware");
+	const stringSchema = builder.primitive("string", ValueSchema.String);
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-example/src/App.tsx:120]: Let's check how `primitive()` works, what that first argument is, and if something like `stringProp`  might be clearer than just ` string` .

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts (hunk 1 above)
Comment (jikim-msft) [reply packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts]: Moved the class to `experimental/dds/tree2/src/shared-tree/sharedTree.ts`.

File: packages/tools/devtools/devtools-example/src/App.tsx [hunk 7]
```diff
@@ -93,6 +112,30 @@ async function populateRootMap(container: IFluidContainer): Promise<void> {
 	const sharedCounter = await container.create(SharedCounter);
 	rootMap.set(sharedCounterKey, sharedCounter.handle);
 
+	// Set up SharedTree for visualization
+	const sharedTreeFactory = SharedTreeFactoryObject as unknown as SharedObjectClass<ISharedTree>;
```
Comment (jikim-msft) [inline packages/tools/devtools/devtools-example/src/App.tsx:116]: Would like to have idea on naming this instance 😅.

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts (hunk 1 above)
Comment (Josmithr) [reply packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts]: I would love to see a larger conversation around this. 100% agreed that we shouldn't be exporting the implementation classes. But we should ensure it remains possible to pattern-match on kinds of DDSs, and that there is an ergonomic way to do so.

This workaround effectively means that consumers lose the ability to provide custom tree visualizers (a degree of freedom they have for the other DDSs), unless they know how to implement this same workaround..

Comment (Josmithr) [reply packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts]: @jikim-msft would it be easier to just inline `new SharedTreeFactory().type` in our visualizer array declaration, rather than adding this type?

File: packages/tools/devtools/devtools-example/src/App.tsx (hunk 5 above)
Comment (Josmithr) [reply packages/tools/devtools/devtools-example/src/App.tsx]: I actually didn't even look at the console statement. I just saw the TODO and figured I would call it out. 😋

File: experimental/dds/tree2/src/shared-tree/sharedTree.ts [hunk 8]
```diff
@@ -241,3 +241,15 @@ export class SharedTreeFactory implements IChannelFactory {
...
 	}
 }
+
+/**
+ * SharedTree class object containing static factory method used for {@link @fluidframework/fluid-static#IFluidContainer}.
+ *
+ * @public
+ */
+// eslint-disable-next-line @typescript-eslint/no-extraneous-class
+export class SharedTreeFactoryObject {
```
Comment (noencke) [inline experimental/dds/tree2/src/shared-tree/sharedTree.ts:251]: This being available here still doesn't solve your use case (or others with the same problem); it doesn't get you the class object you need to satisfy the container create API. For that, you also need the `SharedTreeFactoryObject as unknown as SharedObjectClass<ISharedTree>` cast. If the goal of putting it here is to add this "kludge" to a central location so that others can use the kludge as well, then it should probably be the "full kludge". As it is now, a consumer could get this class, but then they still have to do the cast; at that point, their hands are dirty enough that it wouldn't have been much more work for them to also add these four lines of boilerplate.

I think the most helpful thing would be to put all the kludginess in one place and have it well documented. I would suggest:

* Make this class generic over the type of factory, and have it take a factory when constructed, which is the factory returned by `getFactory()`.
* Don't expose it. Instead, expose a free function which takes a factory, creates an instance of this class, and returns a `SharedObjectClass` (it does the cast on the return value).
* Move the function and class to the fluidstatic package and import from there.
* Document the free function thoroughly as a workaround, add a TODO, etc.

That way, the kludge is well-bounded and communicated. 

Or, maybe we can just change fluid static to take a factory instead of SharedObjectClass?

Comment (Josmithr) [reply experimental/dds/tree2/src/shared-tree/sharedTree.ts:251]: @ssimic2 @ChumpChief Interested in your thoughts here as it pertains to the fluid-static API. In particular, would it make sense to update fluid-static to take in a factory rather than `SharedObjectClass`?

Comment (noencke) [reply experimental/dds/tree2/src/shared-tree/sharedTree.ts:251]: [This comment](https://github.com/microsoft/FluidFramework/pull/15813#discussion_r1214862980) has some additional context that we've discussed with @Abe27342 as well, namely a potential additional motivation for updating the fluid-static APIs going forward.

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts (hunk 2 above)
Comment (jikim-msft) [reply packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts]: `DefaultVisualizer` now uses `iterateNodes` and `iterateFields` for iterating through the tree-data.

File: packages/tools/devtools/devtools-example/src/App.tsx [hunk 9]
```diff
@@ -39,19 +46,46 @@ const sharedTextKey = "shared-text";
...
  * Key in the app's `rootMap` under which the SharedCell object is stored.
  */
 const emojiMatrixKey = "emoji-matrix";
 
+
+/**
+ * SharedTree class object containing static factory method used for {@link @fluidframework/fluid-static#IFluidContainer}.
+ */
+// eslint-disable-next-line @typescript-eslint/no-extraneous-class
+class SharedTree {
```
Comment (noencke) [inline packages/tools/devtools/devtools-example/src/App.tsx:64]: nit: you can define this class inside the body of the `castSharedTreeType` function. Then, the class is not accessible anywhere else and the `castSharedTreeType` function entirely contains the kludge. Bonus: then you don't need to document it ;) (although you should still document the `castSharedTreeType` function so other people know why we need it for now).

File: pnpm-lock.yaml [hunk 10]
```diff
@@ -15435,6 +15435,9 @@ importers:
 
   packages/tools/devtools/devtools-core:
     dependencies:
+      '@fluid-experimental/tree2':
```
Comment (CraigMacomber) [inline pnpm-lock.yaml:15438]: changes to this file seem unrelated.

File: packages/tools/devtools/devtools-core/package.json [hunk 11]
```diff
@@ -59,6 +59,7 @@
 		"temp-directory": "nyc/.nyc_output"
 	},
 	"dependencies": {
+		"@fluid-experimental/tree2": "workspace:~",
```
Comment (CraigMacomber) [inline packages/tools/devtools/devtools-core/package.json:62]: Why does this PR need this?

File: packages/tools/devtools/devtools-example/src/App.tsx (hunk 9 above)
Comment (jikim-msft) [reply packages/tools/devtools/devtools-example/src/App.tsx]: 🐐 feedback. Thank you!

File: packages/tools/devtools/devtools-example/src/App.tsx [hunk 12]
```diff
@@ -39,19 +47,51 @@ const sharedTextKey = "shared-text";
...
+ */
+const sharedTreeKey = "shared-tree";
+
 /**
  * Key in the app's `rootMap` under which the SharedCell object is stored.
  */
 const emojiMatrixKey = "emoji-matrix";
 
+/**
+ * TODO
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/App.tsx:61]: TODO :)

Comment (Josmithr) [reply packages/tools/devtools/devtools-example/src/App.tsx:61]: We should be sure to note that this is a workaround for an API limitation.

**Review by Josmithr — COMMENTED** [review]
This looks good to me! I do think we should add visualization unit tests before we merge, though :)

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts [hunk 13]
```diff
@@ -190,6 +197,45 @@ export const visualizeSharedString: VisualizeSharedObject = async (
...
+export const visualizeSharedTree: VisualizeSharedObject = async (
+	sharedObject: ISharedObject,
+	visualizeChildData: VisualizeChildData,
+): Promise<FluidObjectTreeNode> => {
+	const sharedTree = sharedObject as ISharedTree;
+
+	const contextRoot = sharedTree.context.root;
+	const children: Record<string, VisualChildNode> = {};
+
+	const iterateNodes = async (field: UntypedField): Promise<void> => {
```
Comment (CraigMacomber) [inline packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:212]: This is unneeded. UntypedField is already iterable.

Comment (CraigMacomber) [reply packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:212]: At least the for loop can be replaced with a for of loop like:
```
for (const child of field) {
	
}
```

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts [hunk 14]
```diff
@@ -190,6 +197,45 @@ export const visualizeSharedString: VisualizeSharedObject = async (
...
+		for (let i = 0; i < field.length; i++) {
+			const node = field.getNode(i);
+			await iterateFields(node);
+		}
+	};
+
+	const iterateFields = async (node: UntypedTree): Promise<void> => {
+		for (const fieldName of Object.getOwnPropertyNames(node)) {
+			const field = node[getField](brand(fieldName));
+			const renderedChild = await visualizeChildData(field);
```
Comment (CraigMacomber) [inline packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:222]: UntypedTree is iterable. It would be much simpler to do something like:

```
for (const field of node) {
	
}
```

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts [hunk 15]
```diff
@@ -190,6 +197,45 @@ export const visualizeSharedString: VisualizeSharedObject = async (
...
+			children[field.fieldKey as string] = renderedChild;
+		}
+	};
+
+	if (sharedTree.root !== undefined) {
+		await iterateNodes(contextRoot);
+	}
```
Comment (CraigMacomber) [inline packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:229]: I don't think this `if` is a good idea. If you are going with option 1 from https://github.com/microsoft/FluidFramework/pull/15813/files/61e51960e1b1fe0c5876c784684ab84a019bbd28#r1214898086 (which it seems like you are) then a tree with a root than unwraps to undefined should still be displayed so you can tell what is in the actual tree (It might have a single node at the root that has a primary optional field thats empty, or might just contain no nodes at all for example: both would show the root as undefined).

contextRoot will never be undefined, so you can and should just do:
```suggestion
	await iterateNodes(contextRoot);
```

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts [hunk 16]
```diff
@@ -190,6 +195,44 @@ export const visualizeSharedString: VisualizeSharedObject = async (
 	};
 };
 
+/**
+ * {@link VisualizeSharedObject} for {@link SharedTree}.
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:199]: Nit
```suggestion
 * {@link VisualizeSharedObject} for {@link ISharedTree}.
```

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts [hunk 17]
```diff
@@ -190,6 +195,44 @@ export const visualizeSharedString: VisualizeSharedObject = async (
...
+
+	const iterateFields = async (node: UntypedTree): Promise<void> => {
+		for (const field of node) {
+			const renderedChild = await visualizeChildData(field);
+
+			children[field.fieldKey as string] = renderedChild;
+		}
+	};
+
+	if (sharedTree.root !== undefined) {
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:224]: Just curious: what does it mean for the root of the tree to be `undefined`?

**Review by Josmithr — APPROVED** [review]
Looks good to me! I still think we want to investigate a more SharedTree-centric way of recursing the tree nodes for visualization, but this is a great first step. Thank you!

File: packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts (hunk 17 above)
Comment (jikim-msft) [reply packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:224]: This should've been removed. As @CraigMacomber mentioned in the previous comment. I will apply it right now.
